src/ast.rs

```diff
@@ -446,6 +446,16 @@ pub trait List: Node {
     }
 }
 
+/// This is for optional values and for lists.
+trait CheckParse {
+    /// A true return means we should descend into the production, false means stop.
+    fn can_be_parsed(pop: &mut Populator<'_>) -> bool;
+}
```

Comment:
Really nice work.

Getting rid of virtual dispatch seems like a good direction, a big one would be  https://github.com/fish-shell/fish-shell/pull/9636#issuecomment-1890470676

Reply:
I'll look into that once this step is merged.
I'm quite fond of `enum_dispatch`, but it might not be the best fit here. A custom macro (perhaps with the help of `paste`) would allow us to deduplicate more code (I'm mostly thinking about the downcasting).

Reply:
possibly but we also need to implement generic methods like `Node::typ()` and `Node::source_range()`, ideally without splitting the node definition paragraphs.

> My original goal was to get rid of the None variants on StatementVariant and BlockStatementHeaderVariant, but I discovered that that would require a larger refactor.

One problem is that those union nodes are not real nodes. It's unclear if they should be but the status quo requires a lot of boilerplate that even spreads to macros (`visit_union_field_mut`).
Maybe we should remove union nodes like `Statement` and replace them with something that's not a full node but simply delegates to the concrete statement.